src/extensions/default/SVGCodeHints/main.js

```diff
@@ -0,0 +1,255 @@
+     * 
+     * @param {Editor} editor
+     * An instance of Editor
+     * 
+     * @param {string} implicitChar
+     * A single character that was inserted by the user or null if the request
+     * was explicity made to start hinting session.
+     * 
+     * @return {boolean}
+     * Determines whether or not hints are available in the current context.
+     */
+    SVGCodeHints.prototype.hasHints = function (editor, implicitChar) {
+        if (editor.getModeForSelection() === "image/svg+xml" && isSVGEnabled) {
+            this.editor = editor;
+            this.tagInfo = SVGUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+            
+            if (this.tagInfo && this.tagInfo.tokenType) {
+                return true;
+            }
+        }
+        return false;
+    };
+    
+    /**
+     * Returns a list of hints that are available in the current context,
+     * or null if there are no hints available. 
+     *
+     * @param {string} implicitChar
+     * A character that the user typed in the hinting session.
+     * 
+     * @return {hints: Array.<string>, match: string, selectInitial: boolean, handleWideResults: boolean}
+     */
+    SVGCodeHints.prototype.getHints = function (implicitChar) {
+        var hints = [], query, tagInfo, i, attrs = [], index, isMultiple;
+        
+        tagInfo  = SVGUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+        this.tagInfo = tagInfo;
+        
+        if (tagInfo && tagInfo.tokenType) {
+            query = tagInfo.token.string.substr(0, tagInfo.offset).trim();
+        
+            if (tagInfo.tokenType === SVGUtils.TokenTag) {
+                for (i in tags) {
+                    if (tags.hasOwnProperty(i) && i.indexOf(query) === 0) {
+                        hints.push(i);
+                    }
+                }
+            } else if (tagInfo.tokenType === SVGUtils.TokenAttr) {
+                if (!tags[tagInfo.tagName]) {
+                    return null;
+                }
+                // Get attributes directly or via an alias.
+                attrs = tags[tagInfo.tagName].attributes || tags[tags[tagInfo.tagName].alias].attributes;
+                hints = attrs.filter(function (value) {
+                    if (value.indexOf(query) === 0 && tagInfo.exclusionList.indexOf(value) === -1) {
+                        return true;
+                    }
+                });
+            } else if (tagInfo.tokenType === SVGUtils.TokenValue) {
+                index = tagInfo.tagName + "/" + tagInfo.attrName;
+                if (!tags[tagInfo.tagName] && !(attributes[index] || attributes[tagInfo.attrName])) {
+                    return null;
+                }
+                
+                // Get attribute options.
+                if (attributes[index]) {
+                    // Prefer tag/attribute for specific tags.
+                    attrs = attributes[index].attribOptions || attributes[attributes[index].alias].attribOptions;
```

Comment:
Would be better if you'd store `attributes[index]` in a variable to simplify this line.

(same with `attributes[tagInfo.attrName]` below)